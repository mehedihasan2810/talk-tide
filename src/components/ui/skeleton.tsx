"use client";

import * as React from "react";
import { cn } from "@/lib/utils"

function Skeleton({
  className,
  ...props
}: React.HTMLAttributes<HTMLDivElement>) {
  return (
    <div
      className={cn("animate-pulse rounded-md bg-gray-300", className)}
      {...props}
    />
  );
}

export { Skeleton };
