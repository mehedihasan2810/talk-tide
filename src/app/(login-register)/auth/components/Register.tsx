"use client";

import { zodResolver } from "@hookform/resolvers/zod";
import { useForm } from "react-hook-form";
import * as z from "zod";
import { Button } from "@/components/ui/button";
import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import {
  ArrowPathIcon,
  ExclamationTriangleIcon,
  LockClosedIcon,
  XMarkIcon,
} from "@heroicons/react/20/solid";
import { signIn, useSession } from "next-auth/react";
import { RegisterSchema } from "@/utils/zod-schema/registerSchema";
import { useState } from "react";
import { Alert, AlertDescription, AlertTitle } from "@/components/ui/alert";
import { useRouter } from "next/navigation";
import { useToast } from "@/components/ui/use-toast";

interface Props {
  updateIsLogin(): void;
}

export default function Register({ updateIsLogin }: Props) {
  const [error, setError] = useState<string | null | undefined>(null);
  const [isLoading, setIsLoading] = useState<boolean>(false);

  const router = useRouter();

  const session = useSession();

  const { toast } = useToast();

  // useForm
  const form = useForm<z.infer<typeof RegisterSchema>>({
    resolver: zodResolver(RegisterSchema),
    defaultValues: {
      email: "",
      username: "",
      password: "",
    },
  });

  async function onSubmit(values: z.infer<typeof RegisterSchema>) {
    // if user is already logged in then don't proceed. return with a message
    if (session.data) {
      setError(
        "You are already logged in! If you want to create new account then logout first",
      );
      return;
    }

    setIsLoading(true);

    // log in the user with required credentials
    const res = await signIn("credentials", {
      redirect: false,
      // redirect: true,
      // callbackUrl: "/",
      authType: "register",
      ...values,
    });

    setIsLoading(false);

    // if login is not successful then show a error message
    if (!res?.ok || res.error) {
      setError(res?.error);
    } else {
      //  if the login successful then reset the form and redirect the user to chat page

      toast({ title: "You have registered successfully", variant: "success" });

      form.reset();
      router.replace("/chat");
    }
  }
  return (
    <div className="grid h-screen w-screen place-items-center">
      <Form {...form}>
        <form
          data-testid="register-form"
          onSubmit={form.handleSubmit(onSubmit)}
          className="w-80 space-y-5"
        >
          {/* header starts */}
          <div className="mb-8 flex items-center justify-center gap-1 text-primary">
            <h2 className="scroll-m-20 text-3xl font-semibold tracking-tight">
              Register
            </h2>
            <LockClosedIcon className="h-8 w-8" />
          </div>
          {/* header ends */}

          {/* Alert starts */}
          {error && (
            <Alert variant="destructive" className="relative">
              <button
                onClick={() => setError(null)}
                className="absolute right-2 top-1 rounded-full bg-rose-50 p-1"
                type="button"
              >
                <XMarkIcon className="h-5 w-5" />
              </button>
              <ExclamationTriangleIcon className="h-4 w-4" />
              <AlertTitle>Error!</AlertTitle>
              <AlertDescription>{error}</AlertDescription>
            </Alert>
          )}
          {/* Alert ends */}
          <FormField
            control={form.control}
            name="email"
            render={({ field }) => (
              <FormItem>
                <FormControl>
                  <Input
                    placeholder="Enter your email..."
                    {...field}
                    className="h-11 border-primary/40 text-base focus-visible:ring-primary"
                  />
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />

          <FormField
            control={form.control}
            name="username"
            render={({ field }) => (
              <FormItem>
                <FormControl>
                  <Input
                    placeholder="Enter your username..."
                    {...field}
                    className="h-11 border-primary/40 text-base focus-visible:ring-primary"
                  />
                </FormControl>

                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="password"
            render={({ field }) => (
              <FormItem>
                <FormControl>
                  <Input
                    type="password"
                    placeholder="Enter your password..."
                    {...field}
                    className="h-11 border-primary/40 text-base focus-visible:ring-primary"
                  />
                </FormControl>

                <FormMessage />
                <FormDescription className="text-center">
                  Already have an account?{" "}
                  <button
                    type="button"
                    disabled={isLoading}
                    onClick={updateIsLogin}
                    className="text-primary underline hover:text-primary/70"
                  >
                    Login
                  </button>
                </FormDescription>
              </FormItem>
            )}
          />
          <Button
            disabled={isLoading || session.status === "loading"}
            type="submit"
            className="h-11 w-full text-base"
          >
            {isLoading && (
              <ArrowPathIcon className="mr-2 h-4 w-4 animate-spin" />
            )}{" "}
            Register
          </Button>
        </form>
      </Form>
    </div>
  );
}
