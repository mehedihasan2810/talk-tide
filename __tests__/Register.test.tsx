import Register from "@/app/(login-register)/auth/components/Register";
import { useToast } from "@/components/ui/use-toast";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { signIn, useSession } from "next-auth/react";
import { useRouter } from "next/navigation";

jest.mock("next-auth/react", () => ({
  ...jest.requireActual("next-auth/react"),
  useSession: jest.fn().mockReturnValue({
    status: "unauthenticated",
    data: null,
  }),
  signIn: jest.fn(),
}));

jest.mock("@/components/ui/use-toast", () => ({
  useToast: jest.fn().mockReturnValue({
    toast: jest.fn(),
  }),
}));

jest.mock("next/navigation", () => ({
  useRouter: jest.fn().mockReturnValue({
    refresh: jest.fn(),
    replace: jest.fn(),
  }),
}));

describe("Register Component", () => {
  test("Should render perfectly", () => {
    // ARRANGE
    const { asFragment } = render(<Register updateIsLogin={() => {}} />);

    // ASSERT
    expect(
      screen.getByRole("heading", { level: 2, name: "Register" }),
    ).toBeInTheDocument();
    expect(screen.getByTestId("register-form")).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText("Enter your email..."),
    ).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText("Enter your username..."),
    ).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText("Enter your password..."),
    ).toBeInTheDocument();
    expect(screen.getByTestId("login-link")).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: "Register" }),
    ).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "Register" })).toBeEnabled();
    expect(asFragment()).toMatchSnapshot();
  });

  test("Password input should have type password", () => {
    // ARRANGE
    render(<Register updateIsLogin={() => {}} />);

    // ASSERT
    const password = screen.getByPlaceholderText("Enter your password...");
    expect(password).toHaveAttribute("type", "password");
  });

  test("Input fields should work", async () => {
    // EVENT
    const user = userEvent.setup();

    // ARRANGE
    render(<Register updateIsLogin={() => {}} />);

    // ACT
    const email = screen.getByTestId("register-email");
    const username = screen.getByTestId("register-username");
    const password = screen.getByTestId("register-password");
    await user.type(email, "test@example.com");
    await user.type(username, "testuser");
    await user.type(password, "testpassword");

    // ASSERT
    expect(email).toHaveValue("test@example.com");
    expect(username).toHaveValue("testuser");
    expect(password).toHaveValue("testpassword");
  });

  test("Validation: email, username & password validation should work properly", async () => {
    // EVENT
    const user = userEvent.setup();

    // ARRANGE
    render(<Register updateIsLogin={() => {}} />);

    // ACT
    const email = screen.getByTestId("register-email");
    const username = screen.getByTestId("register-username");
    const password = screen.getByTestId("register-password");
    const submitBtn = screen.getByRole("button", { name: "Register" });
    await user.type(email, "email");
    await user.type(username, "m");
    await user.type(password, "12345");
    await user.click(submitBtn);

    // ASSERT
    expect(screen.getByText("Invalid email address")).toBeInTheDocument();
    expect(
      screen.getByText("Username must be at least 2 characters."),
    ).toBeInTheDocument();
    expect(
      screen.getByText("Password must be at least 6 characters."),
    ).toBeInTheDocument();
  });

  test("Submits the form successfully", async () => {
    //  MOCK
    const { toast } = useToast();
    const { refresh, replace } = useRouter();
    (signIn as jest.Mock).mockResolvedValue({
      ok: true,
    });

    // EVENT
    const user = userEvent.setup();

    render(<Register updateIsLogin={() => {}} />);

    // ACT
    const email = screen.getByTestId("register-email");
    const username = screen.getByTestId("register-username");
    const password = screen.getByTestId("register-password");
    const submitBtn = screen.getByRole("button", { name: "Register" });
    await user.type(email, "test@example.com");
    await user.type(username, "testuser");
    await user.type(password, "testpassword");
    await user.click(submitBtn);

    // ASSERT
    expect(toast).toHaveBeenCalled();
    expect(refresh).toHaveBeenCalled();
    expect(replace).toHaveBeenCalled();
  });

  test("Displays error message on unsuccessful registration", async () => {
    //  MOCK
    (signIn as jest.Mock).mockResolvedValue({
      ok: false,
      error: "Error message",
    });

    // EVENT
    const user = userEvent.setup();

    // ARRANGE
    render(<Register updateIsLogin={() => {}} />);

    // ACT
    const email = screen.getByTestId("register-email");
    const username = screen.getByTestId("register-username");
    const password = screen.getByTestId("register-password");
    const submitBtn = screen.getByRole("button", { name: "Register" });
    await user.type(email, "test@example.com");
    await user.type(username, "testuser");
    await user.type(password, "testpassword");
    await user.click(submitBtn);

    // ASSERT
    expect(screen.getByText("Error!")).toBeInTheDocument();
    expect(screen.getByText("Error message")).toBeInTheDocument();
  });

  test("Displays error/warning message if user is already logged in", async () => {
    // MOCK
    (useSession as jest.Mock).mockReturnValue({
      status: "authenticated",
      data: {
        user: {
          id: "1",
          email: "test@example.com",
        },
      },
    });

    // EVENT
    const user = userEvent.setup();

    // ARRANGE
    render(<Register updateIsLogin={() => {}} />);

    // ACT
    const email = screen.getByTestId("register-email");
    const username = screen.getByTestId("register-username");
    const password = screen.getByTestId("register-password");
    const submitBtn = screen.getByRole("button", { name: "Register" });
    await user.type(email, "test@example.com");
    await user.type(username, "testuser");
    await user.type(password, "testpassword");
    await user.click(submitBtn);

    // ASSERT
    expect(screen.getByText("Error!")).toBeInTheDocument();
    expect(
      screen.getByText(
        "You are already logged in! If you want to create new account then logout first",
      ),
    ).toBeInTheDocument();
  });
});
