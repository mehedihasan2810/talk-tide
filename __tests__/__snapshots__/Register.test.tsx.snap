// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Component Should render perfectly 1`] = `
<DocumentFragment>
  <div
    class="grid h-screen w-screen place-items-center"
  >
    <form
      class="w-80 space-y-5"
      data-testid="register-form"
    >
      <div
        class="mb-8 flex items-center justify-center gap-1 text-primary"
      >
        <h2
          class="scroll-m-20 text-3xl font-semibold tracking-tight"
        >
          Register
        </h2>
        <svg
          aria-hidden="true"
          class="h-8 w-8"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 1a4.5 4.5 0 00-4.5 4.5V9H5a2 2 0 00-2 2v6a2 2 0 002 2h10a2 2 0 002-2v-6a2 2 0 00-2-2h-.5V5.5A4.5 4.5 0 0010 1zm3 8V5.5a3 3 0 10-6 0V9h6z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="space-y-2"
      >
        <input
          aria-describedby=":r0:-form-item-description"
          aria-invalid="false"
          class="flex w-full rounded-md border bg-background px-3 py-2 ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 h-11 border-primary/40 text-base focus-visible:ring-primary"
          data-testid="register-email"
          id=":r0:-form-item"
          name="email"
          placeholder="Enter your email..."
          value=""
        />
      </div>
      <div
        class="space-y-2"
      >
        <input
          aria-describedby=":r1:-form-item-description"
          aria-invalid="false"
          class="flex w-full rounded-md border bg-background px-3 py-2 ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 h-11 border-primary/40 text-base focus-visible:ring-primary"
          data-testid="register-username"
          id=":r1:-form-item"
          name="username"
          placeholder="Enter your username..."
          value=""
        />
      </div>
      <div
        class="space-y-2"
      >
        <input
          aria-describedby=":r2:-form-item-description"
          aria-invalid="false"
          class="flex w-full rounded-md border bg-background px-3 py-2 ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 h-11 border-primary/40 text-base focus-visible:ring-primary"
          data-testid="register-password"
          id=":r2:-form-item"
          name="password"
          placeholder="Enter your password..."
          type="password"
          value=""
        />
        <p
          class="text-sm text-muted-foreground text-center"
          id=":r2:-form-item-description"
        >
          Already have an account? 
          <button
            class="text-primary underline hover:text-primary/70"
            data-testid="login-link"
            type="button"
          >
            Login
          </button>
        </p>
      </div>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-white hover:opacity-80 px-4 py-2 h-11 w-full text-base"
        type="submit"
      >
         Register
      </button>
    </form>
  </div>
</DocumentFragment>
`;
